gear-programs/vft-manager/app/src/services/submit_receipt/mod.rs

```diff
@@ -0,0 +1,188 @@
+        return Err(Error::NotHistoricalProxy);
+    }
+
+    let receipt =
+        ReceiptEnvelope::decode(&mut &receipt_rlp[..]).map_err(|_| Error::NotSupportedEvent)?;
+
+    if !receipt.is_success() {
+        return Err(Error::NotSupportedEvent);
+    }
+
+    // Decode log and check that it is from an allowed address.
+    let (vara_token_id, event) = receipt
+        .logs()
+        .iter()
+        .find_map(|log| {
+            let address = H160::from(log.address.0 .0);
+            let event = abi::ERC20_MANAGER::BridgingRequested::decode_log_data(log, true).ok()?;
+            let eth_token_id = H160::from(event.token.0 .0);
+            let vara_token_id = service
+                .state()
+                .token_map
+                .get_vara_token_id(&eth_token_id)
+                .ok()?;
+
+            (service.erc20_manager_address() == address).then_some((vara_token_id, event))
+        })
+        .ok_or(Error::NotSupportedEvent)?;
+
+    let transactions = transactions_mut();
+    let key = (slot, transaction_index);
+    if transactions.contains(&key) {
+        return Err(Error::AlreadyProcessed);
+    }
+
+    if transactions.len() >= TX_HISTORY_DEPTH
+        && transactions
+            .first()
+            .map(|first| &key < first)
+            .unwrap_or(false)
+    {
+        return Err(Error::TransactionTooOld);
+    }
+
+    let msg_id = gstd::msg::id();
+    let amount = U256::from_little_endian(event.amount.as_le_slice());
+    let receiver = ActorId::from(event.to.0);
+    let supply_type = service.state().token_map.get_supply_type(&vara_token_id)?;
+    let transaction_details = TxDetails {
+        vara_token_id,
+        receiver,
+        amount,
+        token_supply: supply_type,
+    };
+
+    if transactions.len() >= TX_HISTORY_DEPTH {
+        transactions.pop_first();
+    }
+    transactions.insert((slot, transaction_index));
+
+    msg_tracker_mut().insert_message_info(
+        msg_id,
+        MessageStatus::SendingMessageToWithdrawTokens,
+        transaction_details,
+    );
+
+    match supply_type {
+        TokenSupply::Ethereum => {
+            token_operations::mint(vara_token_id, receiver, amount, service.config(), msg_id)
+                .await
+                .expect("Failed to mint tokens");
+        }
+        TokenSupply::Gear => {
+            token_operations::unlock(vara_token_id, receiver, amount, service.config(), msg_id)
+                .await
+                .expect("Failed to unlock tokens");
+        }
+    }
+
+    Ok(())
+}
+
+/// Try to execute failed request again. It can be used to complete funds withdrawal when
+/// the [submit_receipt] execution unexpectedly finished (due to the insufficient gas amount
+/// or some other temporary error) after message to `VFT` program have already been sent
+/// and failed for some reason (in this case Ethereum transaction hash is already marked as processed).
+pub async fn handle_interrupted_transfer<T: ExecContext>(
```

Comment:
discussed, leave notes here.

From my point of view this `msg_tracker` is redundant. The minimum required gas is applied to all operations, so stopping due to lack of gas in critical places is excluded. Thus, only two situations remain: vft did not mint/unlock coins and we were unable to decode vft response. The last question is more related to philosophical reasoning, so we will leave it.

As a result, only the scenario in which the token returned `false` should be taken into account. In this case, you can save information about the processed transaction only if the coins were successfully minted/unlocked. And in another case require that the message be sent to the vft-manager again.

Reply:
Discussed in DM